firedrake/preconditioners/facet_split.py

```diff
@@ -0,0 +1,252 @@
+        from mpi4py import MPI
+        key = (V, W)
+        if key not in self._permutation_cache:
+            indices = get_permutation_map(V, W)
+            if V.comm.allreduce(numpy.all(indices[:-1] <= indices[1:]), MPI.PROD):
+                self._permutation_cache[key] = None
+            else:
+                self._permutation_cache[key] = indices
+        return self._permutation_cache[key]
+
+    def initialize(self, pc):
+
+        from ufl import (InteriorElement, FacetElement, MixedElement, TensorElement, VectorElement,
+                         FiniteElement, EnrichedElement, TensorProductElement, TensorProductCell, HCurl)
+        from firedrake import FunctionSpace, TestFunctions, TrialFunctions
+        from firedrake.assemble import allocate_matrix, TwoFormAssembler
+        from firedrake.solving_utils import _SNESContext
+        from functools import partial
+
+        _, P = pc.getOperators()
+        appctx = self.get_appctx(pc)
+        fcp = appctx.get("form_compiler_parameters")
+
+        if pc.getType() != "python":
+            raise ValueError("Expecting PC type python")
+
+        ctx = dmhooks.get_appctx(pc.getDM())
+        if ctx is None:
+            raise ValueError("No context found.")
+        if not isinstance(ctx, _SNESContext):
+            raise ValueError("Don't know how to get form from %r", ctx)
+
+        prefix = pc.getOptionsPrefix()
+        options_prefix = prefix + self._prefix
+        options = PETSc.Options(options_prefix)
+        mat_type = options.getString("mat_type", "submatrix")
+
+        problem = ctx._problem
+        a = problem.Jp or problem.J
+        V = a.arguments()[-1].function_space()
+        assert len(V) == 1
+
+        # W = V_interior * V_facet
+        scalar_element = V.ufl_element()
+        if isinstance(scalar_element, (TensorElement, VectorElement)):
+            scalar_element = scalar_element._sub_element
+        tensorize = lambda e, shape: TensorElement(e, shape=shape) if shape else e
+
+        def get_facet_element(e):
+            cell = e.cell()
+            if e.sobolev_space() == HCurl and isinstance(cell, TensorProductCell):
+                sub_cells = cell.sub_cells()
+                degree = max(e.degree())
+                variant = e.variant()
+                Qc_elt = FiniteElement("Q", sub_cells[0], degree, variant=variant)
+                Qd_elt = FiniteElement("RTCE", sub_cells[0], degree, variant=variant)
+                Id_elt = FiniteElement("DG", sub_cells[1], degree - 1, variant=variant)
+                Ic_elt = FiniteElement("CG", sub_cells[1], degree, variant=variant)
+                return EnrichedElement(HCurl(TensorProductElement(FacetElement(Qc_elt), Id_elt, cell=cell)),
+                                       HCurl(TensorProductElement(Qd_elt, FacetElement(Ic_elt), cell=cell)),
+                                       HCurl(TensorProductElement(FacetElement(Qd_elt), InteriorElement(Ic_elt), cell=cell)))
```

Comment:
Does "just" making a `RestrictedElement(element, restriction="facet")` not work?

Reply:
This does not work, see here #2713

Reply:
Does the addition of finat/finat#105 now fix this?

Reply:
yes